pkg/collector/process.go

```diff
@@ -361,47 +375,94 @@ func (cp *CollectingProcess) decodeDataSet(dataBuffer *bytes.Buffer, obsDomainID
+	if _, ok := cp.templatesMap[obsDomainID]; !ok {
+		cp.templatesMap[obsDomainID] = make(map[uint16]*template)
 	}
 	elements := make([]*entities.InfoElement, 0)
 	for _, elementWithValue := range elementsWithValue {
 		elements = append(elements, elementWithValue.GetInfoElement())
 	}
-	cp.templatesMap[obsDomainID][templateID] = elements
-	// template lifetime management
+	tpl, ok := cp.templatesMap[obsDomainID][templateID]
+	if !ok {
+		tpl = &template{}
+		cp.templatesMap[obsDomainID][templateID] = tpl
+	}
+	tpl.ies = elements
+	klog.V(4).InfoS("Added template to template map", "obsDomainID", obsDomainID, "templateID", templateID)
+	// Template lifetime management for UDP.
 	if cp.protocol != "udp" {
 		return
 	}
-	// Handle udp template expiration
-	go func() {
-		ticker := time.NewTicker(cp.templateTTL)
-		defer ticker.Stop()
-		select {
-		case <-ticker.C:
+	tpl.expiryTime = cp.clock.Now().Add(cp.templateTTL)
+	if tpl.expiryTimer == nil {
+		tpl.expiryTimer = cp.clock.AfterFunc(cp.templateTTL, func() {
 			klog.Infof("Template with id %d, and obsDomainID %d is expired.", templateID, obsDomainID)
-			cp.deleteTemplate(obsDomainID, templateID)
-			break
+			now := cp.clock.Now()
+			// From the Go documentation:
+			//   For a func-based timer created with AfterFunc(d, f), Reset either
+			//   reschedules when f will run, in which case Reset returns true, or
+			//   schedules f to run again, in which case it returns false. When Reset
+			//   returns false, Reset neither waits for the prior f to complete before
+			//   returning nor does it guarantee that the subsequent goroutine running f
+			//   does not run concurrently with the prior one. If the caller needs to
+			//   know whether the prior execution of f is completed, it must coordinate
+			//   with f explicitly.
+			// In our case, when f executes, we have to verify that the record is indeed
+			// scheduled for deletion by checking expiryTime. We cannot just
+			// automatically delete the template.
+
+			// No reason to try to stop the timer in this case, even though it would be
+			// technically correct, so we pass false for stopTimer.
+			cp.deleteTemplateWithConds(obsDomainID, templateID, false, func(tpl *template) bool {
+				// lock will be held when this executes
+				return !tpl.expiryTime.After(now)
+			})
+		})
+	} else {
+		tpl.expiryTimer.Reset(cp.templateTTL)
+	}
+}
+
+// deleteTemplate returns true iff a template was actually deleted.
+func (cp *CollectingProcess) deleteTemplate(obsDomainID uint32, templateID uint16) bool {
+	return cp.deleteTemplateWithConds(obsDomainID, templateID, true)
+}
+
+// deleteTemplateWithConds returns true iff a template was actually deleted.
+func (cp *CollectingProcess) deleteTemplateWithConds(obsDomainID uint32, templateID uint16, stopTimer bool, condFns ...func(*template) bool) bool {
+	cp.mutex.Lock()
+	defer cp.mutex.Unlock()
+	template, ok := cp.templatesMap[obsDomainID][templateID]
```

Comment:
Wouldn't the templateID be deleted before, and the timer cannot be stopped when we can't find the templateID in templatesMap?

Reply:
which case / code path are you referring to?
the only case where `stopTimer` should be `false` is if the caller knows for certain than the timer has already expired (taking into account that `condFns` should all be true).

Reply:
I mean, when we remove the `templateID` from a `templatesMap`, shouldn't we call the `expiryTimer.Stop()`? otherwise; wouldn't expiryTimer keep running?

Reply:
Could you also remind me where we call `deleteTemplate`? If we call `deleteTemplate` after a template is expired, deleteTemplate will return false and the `expiryTimer. Stop ()` won't be called?

Reply:
> I mean, when we remove the templateID from a templatesMap, shouldn't we call the expiryTimer.Stop()? otherwise; wouldn't expiryTimer keep running?

We do this a few lines below:
```go
	if stopTimer && template.expiryTimer != nil {
		template.expiryTimer.Stop()
	}
```

> Could you also remind me where we call deleteTemplate? If we call deleteTemplate after a template is expired, deleteTemplate will return false and the expiryTimer. Stop () won't be called?

`deleteTemplate` is not called right now actually. I have a follow up patch which deletes templates when TCP connections are closed, but I wanted to split PRs to keep them small. Sorry for the confusion. At the moment we call `deleteTemplateWithConds` when expiring a template. During this call, we do not need to Stop the timer, as we are guaranteed that the timer has already fired (see comments in `addTemplate`).

Reply:
Thanks for the explanation. So stopTimer is mainly for TCP when we want to delete the template before it is expired.